graceful.go

```diff
@@ -1,75 +1,86 @@
 package graceful
 
 import (
+	"context"
 	"os"
 	"os/signal"
 	"sync"
-	"syscall"
 )
 
-// Terminator is the interface that wraps the Terminate method.
-//
-// Terminate instructs the implementing object to shut down,
-// close, disposes, or clean up any associated resources.
-type Terminator interface {
-	Terminate()
-}
-
-// TerminatorFunc defines a type on func() as a convenience
-type TerminatorFunc func()
+type Manager struct {
+	c      Uint64
+	t      Uint64
+	ctx    context.Context
+	cancel context.CancelFunc
 
-// Terminate implements the Terminator interface on TerminatorFunc.
-func (f TerminatorFunc) Terminate() {
-	f()
+	mu        sync.Mutex
+	cancelMap map[uint64]context.CancelFunc
 }
 
-var (
-	tt []Terminator
-	mu = sync.Mutex{}
-)
-
-// In registers one or more Terminator(s) to execute in order.
-func In(f ...Terminator) {
-	mu.Lock()
-	defer mu.Unlock()
-
-	tt = append(tt, f...)
+// New creates a manager with the given context.
+func New() *Manager {
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Manager{ctx: ctx, cancel: cancel, mu: sync.Mutex{}, cancelMap: make(map[uint64]context.CancelFunc)}
 }
 
-var DefaultRunner = RoutineRunner{}
-
-// Go is used to track go func()
-func Go(f func()) {
-	DefaultRunner.Run(f)
+// Go wraps and executes the given function w with a cancellable context.
+func (m *Manager) Go(w func(ctx context.Context)) {
+	m.c.Inc()
+	go func() {
+		ctx, cancel := context.WithCancel(m.ctx)
+		defer func() {
+			cancel()
+			m.c.Dec()
+		}()
+		w(ctx)
+	}()
 }
```

Comment:
If we don't cancel the context here then we have a context leak. However, if we do then child contexts are canceled.